### ISSUE TITLE
CRHM-cli Unexpected Error: Start Time before first Observation

### ISSUE THREAD

**Issue by acebulsk** [srlabUsask/crhmcode#460]
I am running CRHM-cli on Ubuntu 22.04 and hitting an error for a custom .prj and .obs file. Everything works great with the badlake test files.

Description of issue:

CRHM will crash if my .obs file starts at 2022-10-01 01:00 for a start date of 2022-10-01. It will run and report an output if the .obs file starts at 2022-10-01 00:00 or if I change the start date to 2022-10-02. In either case the model output starts at 00:00.

I do not get this issue for the Badlake73_76.obs file which starts at 1973-01-01 01:00. 

Error from the .log file when using my custom files is:

```
Project file: /home/alex/local-usask/analysis/crhm-analysis/prj/fortress_forest_ridge_ac.prj

Observation file: /home/alex/local-usask/analysis/crhm-analysis/obs/ffr_crhm_modelling_obs_hourly.obs
 
Time of model run: 45608.6 2024 11 12. Program Version: 4.7_16
Start Time before first Observation
```

The custom .prj and .obs files are attached here. I compiled CRHM-cli on my laptop following the instructions here on github and I am up to date on the most recent commit.

Logan Fang tried this using CRHM-GUI on his machine and the attached custom .prj and .obs files worked fine so suggests a bug within the CRHM-cli version.

[crhm-problem-files.zip](https://github.com/user-attachments/files/17724042/crhm-problem-files.zip)

**Comment by loganxingfang** [comment]
I also recalled what previous CRHM programmer suggested about the first timestep in obs file: the first timestep should not be middle night, ie. 2021 10 01 00 00. 

I think what CRHM does when it reads an obs file:
first, it can get the start date information from obs file;
then, it determines the time interval and frequency of obs file by comparing consecutive timesteps;
next, it creates the timestamp with start date and time interval.

I have the following screen shots to confirm how Borland CRHM does these steps.

I opened the ffr_crhm_modelling_obs_hourly.obs file in Borland CRHM and focus on one day data: 2021 10 01 to 2021 10 02; see screen shot
![Borland_CRHM_obs_screen shot](https://github.com/user-attachments/assets/8ac7fd56-0788-4759-9ca3-011aad762dba)

I opened the ffr_crhm_modelling_obs_hourly_LF.obs file in Borland CRHM and focus on one day data: 2021 10 01 to 2021 10 02, and this obs file has purposely added data on 2021 10 01 00 00; see screen shot
![Borland_CRHM_obs1_screen shot](https://github.com/user-attachments/assets/331eb756-4f26-4c05-aaed-c9ac7f79268d)

I also exported this one day data in Borland CRHM Export menu, 
see screen shot here for ffr_crhm_modelling_obs_hourly.obs file
![Borland_CRHM_obs_export_screen shot](https://github.com/user-attachments/assets/4d2823f7-522a-4f4b-92c6-717c3859a079)

see screen shot here for ffr_crhm_modelling_obs_hourly_LF.obs file
![Borland_CRHM_obs1_export_screen shot](https://github.com/user-attachments/assets/fe523bb5-d77b-4743-af8f-24fa120ae885)


I also repeated these above in CRHM_GUI:
1. I opened the ffr_crhm_modelling_obs_hourly.obs file
![CRHM_GUI_obs_screen shot](https://github.com/user-attachments/assets/32959f74-9d80-4a29-beb4-90b59a91ba36)
2. I opened the ffr_crhm_modelling_obs_hourly_LF.obs file
![CRHM_GUI_obs1_screen shot](https://github.com/user-attachments/assets/a9c23717-e622-4378-86a9-3d522cab0b44)
3. I exported one day data from  the ffr_crhm_modelling_obs_hourly.obs file Export menu
![CRHM_GUI_obs_export_screen shot](https://github.com/user-attachments/assets/2f62161e-6ee0-4b3e-b2e3-086cf6e48daa)
4. I exported one day data from  the ffr_crhm_modelling_obs_hourly_LF.obs file Export menu
![CRHM_GUI_obs1_export_screen shot](https://github.com/user-attachments/assets/b9eb3bea-5126-4c6c-b906-cf75b12107c4)

As shown, CRHM_GUI does it slightly different for ffr_crhm_modelling_obs_hourly_LF.obs file when the first timestep starts middle night; CRHM_GUI includes data from 2021 10 01 01 00 to 2021 10 02 23 00, i.e. CRHM_GUI considers 47 timesteps data for displaying one day data starting 2021 10 01 to 2021 10 02 from the obs file with middle night as first timestep.

**Comment by acebulsk** [comment]
Quick follow up on this. I've created a workaround by starting the prj file a day after the .obs file starts which prevents the immediate segfault but many of the output vars are erroneous for the first timestep:

![image](https://github.com/user-attachments/assets/8e0a31c8-b3fb-4b2a-92c6-d7cac366e644)

But they do seem to stabilize on the second time step. So something seems off for the model start up.

**Comment by acebulsk** [comment]
This appears to be fixed now after editing an if statement  in `CRHMmain.cpp` which checks the first row datetime variable in the .obs file against the start time in the .prj file. Originally we checked if `DTstartR < FileData->Dt1`. Even if these two print out to be the same, since both vars are `double` they had slightly different precision and thus was erroneously entering this if statement in some edge cases. I changed this to use a threshold value to check for equality first and then to see if we have obs for the .prj start time.
